external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java

```diff
@@ -0,0 +1,566 @@
+    private transient Map<String, TreeSet<BigInteger>> ackedPerShard = new HashMap<>();
+    // sorted failed sequence numbers - needed to figure out what sequence number can be committed
+    private transient Map<String, TreeSet<BigInteger>> failedPerShard = new HashMap<>();
+    // shard iterator corresponding to position in shard for new messages
+    private transient Map<String, String> shardIteratorPerShard = new HashMap<>();
+    // last fetched sequence number corresponding to position in shard
+    private transient Map<String, String> fetchedSequenceNumberPerShard = new HashMap<>();
+    // shard iterator corresponding to position in shard for failed messages
+    private transient Map<KinesisMessageId, String> shardIteratorPerFailedMessage = new HashMap<>();
+    // timestamp to decide when to commit to zk again
+    private transient long lastCommitTime;
+    // boolean to track deactivated state
+    private transient boolean deactivated;
+    private transient AmazonKinesisClient kinesisClient;
+
+    KinesisRecordsManager (Config config) {
+        this.config = config;
+    }
+
+    void initialize (int myTaskIndex, int totalTasks) {
+        deactivated = false;
+        lastCommitTime = System.currentTimeMillis();
+        initializeKinesisClient();
+        initializeCurator();
+        List<Shard> shards = this.getShards();
+        LOG.info("myTaskIndex is " + myTaskIndex);
+        LOG.info("totalTasks is " + totalTasks);
+        int i = myTaskIndex;
+        while (i < shards.size()) {
+            LOG.info("Shard id " + shards.get(i).getShardId() + " assigned to task " + myTaskIndex);
+            toEmitPerShard.put(shards.get(i).getShardId(), new LinkedList<Record>());
+            i += totalTasks;
+        }
+        initializeFetchedSequenceNumbers();
+        refreshShardIteratorsForNewRecords();
+    }
+
+    void next (SpoutOutputCollector collector) {
+        if (shouldCommit()) {
+            commit();
+        }
+        KinesisMessageId failedMessageId = config.getFailedMessageRetryHandler().getNextFailedMessageToRetry();
+        if (failedMessageId  != null) {
+            // if the retry service returns a message that is not in failed set then ignore it. should never happen
+            BigInteger failedSequenceNumber = new BigInteger(failedMessageId.getSequenceNumber());
+            if (failedPerShard.containsKey(failedMessageId.getShardId()) && failedPerShard.get(failedMessageId.getShardId()).contains(failedSequenceNumber)) {
+                if (!failedandFetchedRecords.containsKey(failedMessageId)) {
+                    fetchFailedRecords(failedMessageId);
+                }
+                if (emitFailedRecord(collector, failedMessageId)) {
+                    failedPerShard.get(failedMessageId.getShardId()).remove(failedSequenceNumber);
+                    config.getFailedMessageRetryHandler().failedMessageEmitted(failedMessageId);
+                    return;
```

Comment:
I thought about it but if i try to reduce the return statements it will create more if else statements. If its not a big issue we can keep as is.